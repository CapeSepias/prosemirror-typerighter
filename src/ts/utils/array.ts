export const compact = <TValue>(
  value: TValue | null | undefined | false | ""
): value is TValue => !!value;
