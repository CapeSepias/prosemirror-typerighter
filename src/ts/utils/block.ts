import { Transaction } from "prosemirror-state";
import { IRange, IBlock } from "../interfaces/IMatch";

export const createBlock = (tr: Transaction, range: IRange): IBlock => {
  // Carriage returns are removed by textBetween, but they're one character
  // long, so if we strip them any position beyond them will be incorrectly offset.
  // The final argument of 'textBetween' here adds a newline character to represent
  // a non-text leaf node.
  const text = tr.doc.textBetween(range.from, range.to, undefined, "\n");
  return {
    text,
    ...range,
    id: createBlockId(tr.time, range.from, range.to)
  };
};

export const createBlockId = (time: number, from: number, to: number) =>
  `${time}-from:${from}-to:${to}`;

export const createMatchId = (
  time: number,
  from: number,
  to: number,
  index: number = 0
) => `${createBlockId(time, from, to)}--match-${index}`;
