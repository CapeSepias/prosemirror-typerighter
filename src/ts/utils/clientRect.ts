export const getClientRectIndex = (event: Event): number | undefined => {

    if (!event.target || !(event.target instanceof HTMLElement) || !(event instanceof MouseEvent)) {
        return undefined;
    }

    const rects = event.target.getClientRects();

    if(rects.length < 1) {
        return 0;
    }

    for (var i = 0; i != rects.length; i++) {
        const {left, right, top, bottom} = rects[i];
        const {pageX, pageY} = event;

        if(pageX >= left && pageX <= right && pageY >= top && pageY <= bottom){
            return i;
        }
    }

    return 0;
}