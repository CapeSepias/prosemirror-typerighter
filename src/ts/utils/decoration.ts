import flatten from "lodash/flatten";
import { Node } from "prosemirror-model";
import { Decoration, DecorationSet } from "prosemirror-view";
import { IRange, IMatches } from "../interfaces/IValidation";

// Our decoration types.
export const DECORATION_VALIDATION = "DECORATION_VALIDATION";
export const DECORATION_VALIDATION_IS_SELECTED =
  "DECORATION_VALIDATION_IS_HOVERING";
export const DECORATION_VALIDATION_HEIGHT_MARKER =
  "DECORATION_VALIDATION_HEIGHT_MARKER";
export const DECORATION_DIRTY = "DECORATION_DIRTY";
export const DECORATION_INFLIGHT = "DECORATION_INFLIGHT";

export const DecorationClassMap = {
  [DECORATION_DIRTY]: "ValidationDebugDirty",
  [DECORATION_INFLIGHT]: "ValidationDebugInflight",
  [DECORATION_VALIDATION]: "ValidationDecoration",
  [DECORATION_VALIDATION_HEIGHT_MARKER]: "ValidationDecoration__height-marker",
  [DECORATION_VALIDATION_IS_SELECTED]: "ValidationDecoration--is-selected"
};

export const DECORATION_ATTRIBUTE_ID = "data-validation-id";
export const DECORATION_ATTRIBUTE_HEIGHT_MARKER_ID = "data-height-marker-id";

export const createDebugDecorationFromRange = (range: IRange, dirty = true) => {
  const type = dirty ? DECORATION_DIRTY : DECORATION_INFLIGHT;
  return Decoration.inline(
    range.from,
    range.to + 1,
    {
      class: DecorationClassMap[type]
    },
    {
      type
    } as any
  );
};

/**
 * Remove decorations from the given ranges. If decorations are found,
 * expand the search range to include their ranges, too.
 */
export const removeDecorationsFromRanges = (
  decorationSet: DecorationSet,
  ranges: IRange[],
  types = [DECORATION_VALIDATION, DECORATION_VALIDATION_HEIGHT_MARKER]
) =>
  ranges.reduce((acc, range) => {
    const predicate = (spec: { [key: string]: any }) =>
      types.indexOf(spec.type) !== -1;
    const decorations = decorationSet.find(range.from, range.to, predicate);
    if (!decorations.length) {
      return acc;
    }
    // Expand the range out to the range of the found decorations.
    const decorationsToRemove = flatten(
      decorations
        .map(decoration =>
          decorationSet.find(decoration.from, decoration.to, predicate)
        )
        .filter(_ => !!_)
    );
    return acc.remove(decorationsToRemove);
  }, decorationSet);

/**
 * Given a validation response and the current decoration set,
 * returns a new decoration set containing the new validations.
 */
export const getNewDecorationsForCurrentValidations = (
  outputs: IMatches[],
  decorationSet: DecorationSet,
  doc: Node
) => {
  // There are new validations available; apply them to the document.
  const decorationsToAdd = createDecorationsForValidationRanges(outputs);

  // Finally, we add the existing decorations to this new map.
  return decorationSet.add(doc, decorationsToAdd);
};

/**
 * Create a height marker element. Used to determine the height
 * of a single line of inline content, which is useful when we're
 * calculating where to place tooltips as the user hovers over multi-
 * line spans.
 */
const createHeightMarkerElement = (id: string) => {
  const element = document.createElement("span");
  element.setAttribute(DECORATION_ATTRIBUTE_HEIGHT_MARKER_ID, id);
  element.className = DecorationClassMap[DECORATION_VALIDATION_HEIGHT_MARKER];
  return element;
};

/**
 * Create a validation decoration for the given range.
 */
export const createDecorationForValidationRange = (
  output: IMatches,
  isSelected = false,
  addHeightMarker = true
) => {
  const className = isSelected
    ? `${DecorationClassMap[DECORATION_VALIDATION]} ${
        DecorationClassMap[DECORATION_VALIDATION_IS_SELECTED]
      }`
    : DecorationClassMap[DECORATION_VALIDATION];
  const opacity = isSelected ? "30" : "07";
  const style = `background-color: #${
    output.category.colour
  }${opacity}; border-bottom: 2px solid #${output.category.colour}`;

  const decorationArray = [
    Decoration.inline(
      output.from,
      output.to,
      {
        class: className,
        style,
        [DECORATION_ATTRIBUTE_ID]: output.matchId
      } as any,
      {
        type: DECORATION_VALIDATION,
        id: output.matchId,
        categoryId: output.category.id,
        inclusiveStart: true
      } as any
    )
  ];

  return addHeightMarker
    ? [
        ...decorationArray,
        Decoration.widget(
          output.from,
          createHeightMarkerElement(output.matchId),
          {
            type: DECORATION_VALIDATION_HEIGHT_MARKER,
            id: output.matchId,
            categoryId: output.category.id
          } as any
        )
      ]
    : decorationArray;
};

export const createDecorationsForValidationRanges = (ranges: IMatches[]) =>
  flatten(ranges.map(_ => createDecorationForValidationRange(_)));

export const findSingleDecoration = (
  decorationSet: DecorationSet,
  predicate: (spec: any) => boolean
): Decoration | undefined => {
  const decorations = decorationSet.find(undefined, undefined, predicate);
  if (!decorations[0]) {
    return undefined;
  }
  return decorations[0];
};
