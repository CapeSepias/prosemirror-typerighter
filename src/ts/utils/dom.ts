import { IStateHoverInfo } from "../state/state";

/**
 * Find the first ancestor node of the given node that matches the selector.
 */
export function findAncestor(
  element: HTMLElement,
  selector: (e: HTMLElement) => boolean
) {
  // tslint:disable-next-line prefer-const
  let currentElement: HTMLElement | null = element;
  while (
    // tslint:disable-next-line no-conditional-assignment
    (currentElement = currentElement.parentElement) &&
    !selector(currentElement)
    // tslint:disable-next-line no-empty
  ) {}
  return currentElement;
}

/**
 * Get the dimensions required for our UI code to render a tooltip. We encapsulate this here
 * to avoid dealing with side effects in the plugin reducer.
 */
export function getStateHoverInfoFromEvent(
  event: MouseEvent,
  containerElement: Element | null,
  heightMarkerElement: Element | null
): IStateHoverInfo | undefined {
  if (
    !event.target ||
    !(event.target instanceof HTMLElement) ||
    !containerElement ||
    !(containerElement instanceof HTMLElement) ||
    !heightMarkerElement ||
    !(heightMarkerElement instanceof HTMLElement)
  ) {
    return;
  }
  const {
    left: elementLeft,
    top: elementTop
  } = event.target.getBoundingClientRect();
  const {
    left: containerLeft,
    top: containerTop
  } = containerElement.getBoundingClientRect();
  const mouseOffsetX = event.clientX - elementLeft;
  const mouseOffsetY = event.clientY - elementTop;
  const { offsetLeft, offsetTop, offsetHeight: height } = event.target;
  return {
    left: elementLeft - containerLeft,
    top: elementTop - containerTop,
    offsetLeft,
    offsetTop,
    height,
    mouseOffsetX,
    mouseOffsetY,
    heightOfSingleLine: heightMarkerElement.offsetHeight
  };
}
