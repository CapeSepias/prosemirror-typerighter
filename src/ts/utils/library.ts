import chunk from "lodash/chunk";
import { MarkTypes } from "./prosemirror";
import { Range } from "..";
import { ValidationLibrary, Operations } from "../interfaces/Validation";

// A temporary validation library
const withoutIndex = <T>(arr: Array<T>, index: number) =>
  arr.slice(0, index).concat(arr.slice(index + 1));

const permutations: <T>(seq: Array<T>) => T[][] = seq =>
  seq.reduce((acc, el, index, arr) => {
    if (!arr.length) return [[]];
    if (arr.length === 1) return [arr];
    return [
      ...acc,
      ...permutations(withoutIndex(arr, index)).map(perms => [el, ...perms], [])
    ];
  }, []);

export const validationLibrary: ValidationLibrary = chunk(
  permutations(Array.from("qwer")).map(perm => {
    const str = perm.join("");
    return {
      regExp: new RegExp(str, "g"),
      annotation: `You used the word ${str}`,
      operation: Operations.ANNOTATE,
      type: MarkTypes.legal
    };
  }),
  500
);
