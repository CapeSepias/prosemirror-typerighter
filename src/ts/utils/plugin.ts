import { stopHoverCommand, stopHighlightCommand } from "../commands";
import { EditorView } from "prosemirror-view";
import { PluginKey } from "prosemirror-state";

export const pluginKey = new PluginKey("prosemirror-typerighter");

export const maybeResetHoverStates = (view: EditorView, ignoreElement: (e: HTMLElement) => boolean, event: Event) => {
  // If we're leaving the editor node to mouse over something that we consider part of
  // the prosemirror-typerighter UI, ignore this event â€“ we're still within relevant UI,
  // and shouldn't reset our highlight or hover states.
  if (
    event instanceof MouseEvent &&
    event.relatedTarget instanceof HTMLElement &&
    ignoreElement(event.relatedTarget)
  ) {
    return false;
  }

  const pluginState = pluginKey.getState(view.state);
  if (pluginState.hoverId) {
    return stopHoverCommand()(view.state, view.dispatch);
  }
  if (pluginState.highlightId) {
    return stopHighlightCommand()(view.state, view.dispatch);
  }
}
