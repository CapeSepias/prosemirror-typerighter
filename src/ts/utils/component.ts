import { IMatch, TyperighterTelemetryAdapter } from "..";
import { maybeGetDecorationElement } from "../utils/decoration";

export const createScrollToRangeHandler = (match: IMatch, getScrollOffset: () => number, editorScrollElement: Element, telemetryAdapter?: TyperighterTelemetryAdapter) => (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
e.preventDefault();
e.stopPropagation();

telemetryAdapter?.sidebarMatchClicked(match, document.URL);

if (!editorScrollElement) {
    return;
}

const decorationElement = maybeGetDecorationElement(match.matchId);

if (decorationElement) {
    const scrollToYCoord = decorationElement.offsetTop - getScrollOffset();
    editorScrollElement.scrollTo({
    top: scrollToYCoord,
    behavior: "smooth"
    });
}
};