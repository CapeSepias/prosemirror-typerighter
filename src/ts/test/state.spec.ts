import { builders } from "prosemirror-test-builder";
import { nodes, marks } from "prosemirror-schema-basic";
import { Transaction } from "prosemirror-state";
import { Schema } from "prosemirror-model";
import {
  validationPluginReducer,
  validationRequestError,
  validationRequestPending,
  validationRequestStart,
  validationRequestSuccess,
  newHoverIdReceived,
  selectValidationById
} from "../state";
import { DecorationSet } from "prosemirror-view";
import { createDebugDecorationFromRange } from "../utils/decoration";
import { PluginState } from "..";

jest.mock("uuid/v4", () => () => "uuid");

const noteSchema = new Schema({
  nodes,
  marks
});

const build = builders(noteSchema, {
  p: {
    markType: "paragraph"
  }
});

const { doc, p } = build;

const initialDocToValidate = doc(p("Example text to validate"));
const initialTr = new Transaction(initialDocToValidate);
initialTr.doc = initialDocToValidate;
initialTr.time = 0;
const initialState: PluginState = {
  currentThrottle: 100,
  initialThrottle: 100,
  maxThrottle: 1000,
  decorations: DecorationSet.create(doc, []),
  dirtiedRanges: [],
  currentValidations: [],
  lastValidationTime: 0,
  hoverId: undefined,
  hoverRect: undefined,
  trHistory: [initialTr],
  validationInFlight: undefined,
  validationPending: false,
  error: undefined
};

describe("State management", () => {
  describe("Action handlers", () => {
    describe("validationRequestPending", () => {
      it("should mark the state as pending validation", () => {
        expect(
          validationPluginReducer(
            initialTr,
            initialState,
            validationRequestPending()
          )
        ).toEqual({
          ...initialState,
          validationPending: true
        });
      });
    });
    describe("validationRequestStart", () => {
      it("should remove the pending status and any dirtied ranges, and mark the validation as in flight", () => {
        const docToValidate = doc(p("Example text to validate"));
        const tr = new Transaction(docToValidate);
        tr.doc = docToValidate;
        tr.time = 1337;
        expect(
          validationPluginReducer(
            tr,
            {
              ...initialState,
              dirtiedRanges: [{ from: 5, to: 10 }],
              validationPending: true
            },
            validationRequestStart()
          )
        ).toEqual({
          ...initialState,
          dirtiedRanges: [],
          decorations: new DecorationSet().add(docToValidate, [
            createDebugDecorationFromRange({ from: 1, to: 25 }, false)
          ]),
          validationPending: false,
          validationInFlight: {
            validationInputs: [
              {
                str: "Example text to validate",
                from: 1,
                to: 25
              }
            ],
            id: 1337
          }
        });
      });
      it("should remove debug decorations, if any", () => {
        const docToValidate = doc(p("Example text to validate"));
        const tr = new Transaction(docToValidate);
        tr.doc = docToValidate;
        tr.time = 1337;
        expect(
          validationPluginReducer(
            tr,
            {
              ...initialState,
              dirtiedRanges: [{ from: 5, to: 10 }],
              decorations: new DecorationSet().add(docToValidate, [
                createDebugDecorationFromRange({ from: 1, to: 3 })
              ]),
              validationPending: true
            },
            validationRequestStart()
          )
        ).toEqual({
          ...initialState,
          dirtiedRanges: [],
          decorations: new DecorationSet().add(docToValidate, [
            createDebugDecorationFromRange({ from: 1, to: 25 }, false)
          ]),
          validationPending: false,
          validationInFlight: {
            validationInputs: [
              {
                str: "Example text to validate",
                from: 1,
                to: 25
              }
            ],
            id: 1337
          }
        });
      });
    });
    describe("validationRequestSuccess", () => {
      it("shouldn't do anything if there's nothing in the response", () => {
        expect(
          validationPluginReducer(
            initialTr,
            initialState,
            validationRequestSuccess({
              validationOutputs: [],
              id: "1337"
            })
          )
        ).toEqual(initialState);
      });
      it("should create decorations for the incoming validations", () => {
        const docToValidate = doc(p("Example text to validate"));
        const tr = new Transaction(docToValidate);
        tr.doc = docToValidate;
        tr.time = 1337;
        expect(
          validationPluginReducer(
            tr,
            initialState,
            validationRequestSuccess({
              validationOutputs: [
                {
                  id: "id",
                  str: "Example text to validate",
                  from: 5,
                  to: 10,
                  annotation: "Summat ain't right",
                  type: "EXAMPLE_TYPE"
                }
              ],
              id: "1337"
            })
          )
        ).toMatchSnapshot();
      });
    });
    describe("validationRequestError", () => {
      it("Should re-add the in-flight validation ranges as dirty ranges, and remove the inflight validation", () => {
        expect(
          validationPluginReducer(
            initialTr,
            {
              ...initialState,
              validationInFlight: {
                validationInputs: [
                  {
                    str: "Example text to validate",
                    from: 1,
                    to: 25
                  }
                ],
                id: 1337
              }
            },
            validationRequestError({
              validationInput: {
                str: "Example text to validate",
                from: 1,
                to: 25
              },
              id: 1337,
              message: "Too many requests"
            })
          )
        ).toEqual({
          ...initialState,
          dirtiedRanges: [
            {
              from: 1,
              to: 25
            }
          ],
          decorations: new DecorationSet().add(initialDocToValidate, [
            createDebugDecorationFromRange({ from: 1, to: 25 })
          ]),
          error: "Too many requests",
          validationInFlight: undefined
        });
      });
    });
    describe("newHoverIdReceived", () => {
      it("should update the hover id", () => {
        const rect = {
          x: 0,
          y: 0,
          bottom: 0,
          height: 0,
          left: 0,
          right: 0,
          top: 0,
          width: 0
        };
        expect(
          validationPluginReducer(
            new Transaction(doc),
            initialState,
            newHoverIdReceived("exampleHoverId", rect)
          )
        ).toEqual({
          ...initialState,
          hoverId: "exampleHoverId",
          hoverRect: rect
        });
      });
    });
  });
  describe("selectors", () => {
    describe("selectValidationById", () => {
      it("should find the given validation by id", () => {
        expect(
          selectValidationById(
            {
              currentValidations: [
                {
                  id: "1"
                },
                {
                  id: "2"
                }
              ]
            } as any,
            "1"
          )
        ).toEqual({ id: "1" });
      });
      it("should return undefined if there is no validation", () => {
        expect(
          selectValidationById(
            {
              currentValidations: [
                {
                  id: "1"
                },
                {
                  id: "2"
                }
              ]
            } as any,
            "3"
          )
        ).toEqual(undefined);
      });
    });
  });
});
