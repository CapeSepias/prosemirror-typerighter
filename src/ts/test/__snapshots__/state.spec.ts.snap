// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State management Action handlers validationRequestSuccess should create decorations for the incoming validations 1`] = `
Object {
  "currentThrottle": 100,
  "currentValidations": Array [],
  "debug": false,
  "decorations": DecorationSet {
    "children": Array [],
    "local": Array [],
  },
  "dirtiedRanges": Array [],
  "error": undefined,
  "hoverId": undefined,
  "hoverInfo": undefined,
  "initialThrottle": 100,
  "maxThrottle": 1000,
  "selectedValidation": undefined,
  "trHistory": Array [
    Transaction {
      "curSelection": undefined,
      "curSelectionFor": 0,
      "doc": Object {
        "content": Array [
          Object {
            "content": Array [
              Object {
                "text": "Example text to validate",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "doc",
      },
      "docs": Array [],
      "mapping": Mapping {
        "from": 0,
        "maps": Array [],
        "mirror": undefined,
        "to": 0,
      },
      "meta": Object {},
      "steps": Array [],
      "storedMarks": undefined,
      "time": 0,
      "updated": 0,
    },
  ],
  "validationPending": false,
  "validationsInFlight": Array [],
}
`;
